### ISSUE TITLE
Precendence of component aliases

### ISSUE THREAD

**Issue by ben-willis** [OpenCageData/address-formatting#58]
I'm trying to write a formatter that passes the test cases but I've run in to a problem with using component aliases.

In this test case we expect the "city_district" to be used as the neighbourhood ahead of "suburb" and substituted in to the formatted address:
```
description: Supermarket, 33.3559,44.4015
components:
    city: Baghdad
    city_district: Rusafa
    country: Iraq
    country_code: iq
    county: Al Resafa
    postcode: 222
    road: A86/N11/D383
    state: Baghdad
    suburb: Mustansiriya
    supermarket: al mustansriya Central Market
expected: |
    al mustansriya Central Market
    Rusafa
    A86/N11/D383
    Baghdad
    222
    Iraq
```
https://github.com/OpenCageData/address-formatting/blob/master/testcases/countries/iq.yaml#L3

However in this case we expect "road_reference" instead of "road_reference_intl" to be used for the road:
```
description: road_reference, 48.406168,2.55452
components:
            country : France
            country_code : fr
            county : Fontainebleau
            postcode : 77630
            road_reference : A 6
            road_reference_intl : E 15
            state : Ile-de-France
            village : Arbonne-la-Forêt
expected:  | 
    A 6
    77630 Arbonne-la-Forêt
    France
```
https://github.com/OpenCageData/address-formatting/blob/master/testcases/countries/fr.yaml#L170

When looking at the components yaml it doesn't look like order is used (city_district comes below suburb under neighbourhood but road_reference comes above road_reference_intl under road):
```
name: road
aliases:
    - footway
    - street
    - street_name
    - residential
    - path
    - pedestrian
    - road_reference
    - road_reference_intl    
---
name: neighbourhood
aliases:
    - suburb
    - city_district
    - district
    - quarter
    - houses
    - subdivision
```
https://github.com/OpenCageData/address-formatting/blob/master/conf/components.yaml#L11

So what order of precedence for component aliases should I use to pass both these test cases?

**Comment by ben-willis** [comment]
My guess is that the existing processers use the ordering of the components to decide which one to use. However my yaml parser loses this ordering and, either way, ordering of keys shouldn't be relied on: https://yaml.org/spec/1.2/spec.html#id2765608

**Comment by freyfogle** [comment]
Hi Ben,

At OpenCage the processor we rely on is [Geo::Address::Formatter](https://github.com/OpenCageData/perl-Geo-Address-Formatter), written in perl. Even if you aren't a perl programmer I suspect you can follow the logic there.

Yes, it does rely on the ordering of the components. I suppose we could change the configuration to define an order explicitly. ie `order = 1`, `order = 2`, etc.

Regarding your question though, bear in mind that while there is a template defined for each country (ISO code), if not enough of the components exist - for example we have just a city and state - then we drop back to the fallback template (of which there are also a few defined). 

I hope that helps clarify things. Great to see you getting involved with the project. Can you share your exact use case?

**Comment by ben-willis** [comment]
Hi Ed,

Yeah I've been trying to follow the logic in the existing processers and have got it down to just one test case failing (the French address). I guess I'm going to have to go back and see if there's some way I can preserve the order from the YAML for the time being.

Going forward is there any reason we can't use the order of the aliases in components.yaml to determine precedence rather than the order of the components in the address?

Yeah that does help, thanks for your quick reply :) As for my use case I'm attempting to write a processor in Scala, initially as a bit of exercise but with the potential to be used in an ETL pipeline using Spark.

**Comment by JirkaChadima** [comment]
TBH this is quite confusing :-) I'd expect the formatter to be input-agnostic, i.e. not to rely on the order of the input, but on the order of the formatter's declaration.

**Comment by freyfogle** [comment]
Hi everyone, I think we are being imprecise in our wording and thus confusing ourselves.

Yes, the processor we use does depend on the ordering in the YAML configuration. If we have both a `road_reference` and a `road_reference_intl` the `road_reference` will be used.

I am happy to help make this more clear, but right now I am on holiday with my kids for next two weeks, so ideally it can wait until after that.

**Comment by ben-willis** [comment]
Happy that this particular test case has been resolved with #62
